### PR TITLE
feat: add academy progress tracker contract

### PR DESCRIPTION
**Pull request by respp**
# Academy Progress Tracker Contract Implementation

## Overview
Implementation of a smart contract to track student progress in an academy setting. The contract allows tracking lesson completion by user and chapter, calculating completion percentages, and emitting events for progress tracking.

## Features Implemented
- **Lesson Tracking**: Track completion of individual lessons by user and chapter
- **Chapter Progress**: Calculate and track chapter completion percentage
- **Event System**: Emit events for both lesson and chapter completion
- **Duplicate Prevention**: Prevent marking the same lesson as completed multiple times
- **Progress Queries**: Functions to query completed lessons and chapter progress

## Technical Details
- **Contract Structure**:
  - `lib.rs`: Main contract implementation
  - `events.rs`: Event definitions and handling
  - `errors.rs`: Custom error types and handling
  - `testutils.rs`: Testing utilities
  - `test.rs`: Unit tests

- **Key Functions**:
  - `mark_lesson_complete`: Mark a lesson as completed
  - `get_completed_lessons`: Get list of completed lessons
  - `get_chapter_completion_percent`: Calculate chapter completion percentage
  - `is_chapter_complete`: Check if chapter is fully completed
  - `set_chapter_lessons`: Configure total lessons in a chapter

- **Events**:
  - `lesson_completed`: Emitted when a lesson is completed
  - `chapter_completed`: Emitted when all lessons in a chapter are completed

## Testing
- Unit tests implemented for all major functionality
- Test coverage for:
  - Lesson completion tracking
  - Chapter progress calculation
  - Event emission
  - Error handling
  - Edge cases (duplicate completions, invalid inputs)

## Documentation
- README.md with:
  - Contract description
  - Building instructions
  - Testing instructions
  - Function documentation
  - Event documentation

## Dependencies
- soroban-sdk
- testutils for testing

## Build Instructions
```bash
stellar contract build --manifest-path ./contracts/academy-progress-tracker/Cargo.toml
```
or
```bash
cd contracts/academy-progress-tracker
stellar contract build
```

## Testing Instructions
```bash
cargo test --features testutils
```
image

## Summary by CodeRabbit

- **New Features**
  - Introduced the Academy Progress Tracker smart contract for tracking student lesson and chapter completions.
  - Users can mark lessons as complete, view completed lessons, check chapter completion percentages, and verify chapter completion status.
  - Events are emitted for lesson and chapter completions.
  - Error handling for invalid operations such as duplicate completions or unauthorized actions.

- **Documentation**
  - Added a comprehensive README detailing contract usage, available functions, event types, and error descriptions.

- **Tests**
  - Added tests to verify progress tracking and event emission.
  - Included test utilities for easier contract testing and interaction.

### CODE REVIEW

**Review by Bran18 — APPROVED** [review]
Hey Renzo, 

Great job on one of your first contracts with Soroban! Your performance was impressive, and I noticed your attention to detail. It seems you spent a few hours studying the Stellar documentation and learning how to use Soroban effectively, as I see you're approaching things in the right way.

Your performance was excellent, and it's clear that you’re following the documentation well. Let's keep building contracts together using Soroban!

I'm merging this one.

File: apps/contract/contracts/academy-progress-tracker/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            .unwrap_or(0);
+
+        completed_lessons.len() as u32 == total_lessons
+    }
+
+    pub fn set_chapter_lessons(env: Env, chapter_id: u32, total_lessons: u32) {
+        env.storage()
+            .persistent()
+            .set(&DataKey::ChapterLessons(chapter_id), &total_lessons);
+    }
```
[the remark this answers is not in the document]
Comment (AndlerRL) [reply apps/contract/contracts/academy-progress-tracker/src/lib.rs:127]: @coderabbitai Let's track this into a new issue to improve this section.

**Review by AndlerRL — APPROVED** [review]
As previously reviewd by @Bran18, everything looks good to go! I made a final sync with develop before the merge :ok_hand: Thanks for your efforts! We see you in another one! :rocket:
